File: Problems/Problem 4/C/problem_4.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
/**
 * Problem 4 from Project Euler
 *
 * A palindromic number read the same in both ways. The largest palindrome made from the proctur of two 2-digit numbers is 9009.
 * Find the largest palindrome from the product of two 3-digit numbers
 *
 * PLEAE NOTE
 * To compile this algorithm you should have gcc, or similar, installed.
 *
 * REGARDING THE PERFORMANCE
 * This is not the final version of the algorithm, consider this a start. If you have a better version please submit a pull request
 * or open an issue on Github.
 *
 * @author Claudio Ludovico Panetta (@Ludo237)
 * @version 1.0.0
 **/

/*
 * reverse
 * @param int number
 * @return int
 * @description
 *  Since we cannot treath number as string in a strict typed language
 *  we have to find another way to solve the palindrome check
 */
int reverse(int number);

int main(int argc, char **argv)
{
    printf("Algorithm start, please wait...\n");
    // We need to store the palindrome, the size of our test and a temporary variable
    // for our loops
    int palindrome = 0, size = (pow(10,3) - 1), solver = 0;
    
    for( int i = size; i > 0; i--)
    {
        for(int j = size; j > 0; j--)
        {
            solver = i * j; // Store the temporary multiplication
            if( (solver > palindrome) && (reverse(solver) == solver) )
            {
                palindrome = solver; // Found the actual palindrome...
            }
        }
    }

    printf("Algorithm finish.");
    return 0;
}

int reverse(int number)
{
    int temp = number, reverse; // Variable support
    while(temp != 0)
    {
        reverse = (reverse * 10) + (temp % 10) ;
        temp /= 10;
    }
    return reverse;
}

File: Problems/Problem 4/PHP/problem_4.php
<?php
/**
 *
 * Problem 4 from Project Euler
 *
 * A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
 * Find the largest palindrome made from the product of two 3-digit numbers.
 *
 *
 * PLEASE NOTE
 * In order to test this script you must have PHP 5.3+ installed and php-cli, to test it with a terminal.
 *
 * REGARDING THE PERFORMANCE
 * This is not the final version, as you can see I'll trace the version of every solution, so please don't think that
 * this is the best algorithm in terms of speed and code optimization, for any trouble open an issue.
 *
 * @author Pietro Arturo PAnetta (@arturu)
 * @author Santo Antonio Terranova (@tng46)
 * @version 1.0
 */

// Starting the benchmark
$startTime = microtime();

// thanks to Santo Antonio Terranova @tng46
function isPalindrome($num){

	if ( strcmp($num, strrev($num)) == 0 )  
            return true;                        //..is palindrome!

        return false;
}

function cycleManager($coppia,$command){

	if ( $command == 'break' ){
		--$coppia['a'];
		$coppia['b']=999;
	}
	else
		--$coppia['b'];

	if ( $coppia['a'] < 100 )
		$coppia['cycle']=false;

	return $coppia;
}


function maxPalindrome($coppia){

	$maxPalindrome = 0;

	$coppia['cycle']=true;

	do {

		$number = $coppia['a'] * $coppia['b'];

		if ( isPalindrome( $number ) ) {
			$maxPalindrome = ( $number > $maxPalindrome ) ? $number : $maxPalindrome;
			$coppia = cycleManager($coppia,'break');
		}

		else 
			$coppia = cycleManager($coppia,'continue');

	} while ($coppia['cycle']);
	

	return $maxPalindrome;
}


$result = maxPalindrome( array('a'=>999,'b'=>999) );

// Finish the benchmark 
$endTime = microtime();
echo "Algorithm time: ". ($endTime - $startTime);
